askama/src/filters/json.rs

```diff
@@ -17,10 +18,58 @@ use serde_json::to_writer_pretty;
 /// To use it in HTML attributes, you can either use it in quotation marks `"{{data|json}}"` as is,
 /// or in apostrophes with the (optional) safe filter `'{{data|json|safe}}'`.
 /// In HTML texts the output of e.g. `<pre>{{data|json|safe}}</pre>` is safe, too.
-pub fn json<S: Serialize>(s: S) -> Result<String> {
-    let mut writer = JsonEscapeBuffer::new();
-    to_writer_pretty(&mut writer, &s).map_err(Error::from)?;
-    Ok(writer.finish())
+#[inline]
+pub fn json<S: Serialize>(s: S) -> Result<ToJson<S>, Infallible> {
+    Ok(ToJson(s))
+}
+
+#[derive(Debug, Clone)]
+pub struct ToJson<S: Serialize>(S);
+
+impl<S: Serialize> fmt::Display for ToJson<S> {
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        to_writer_pretty(JsonWriter(f), &self.0).map_err(|_| fmt::Error)
+    }
+}
+
+struct JsonWriter<'a, 'b: 'a>(&'a mut fmt::Formatter<'b>);
+
+impl io::Write for JsonWriter<'_, '_> {
+    #[inline]
+    fn write(&mut self, bytes: &[u8]) -> io::Result<usize> {
+        self.write_all(bytes)?;
+        Ok(bytes.len())
+    }
+
+    #[inline]
+    fn write_all(&mut self, bytes: &[u8]) -> io::Result<()> {
+        write(self.0, bytes).map_err(|err| io::Error::new(io::ErrorKind::InvalidData, err))
+    }
+
+    #[inline]
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+fn write(f: &mut fmt::Formatter<'_>, bytes: &[u8]) -> fmt::Result {
+    let mut last = 0;
+    for (index, byte) in bytes.iter().enumerate() {
+        let escaped = match byte {
+            b'&' => Some(br"\u0026"),
+            b'\'' => Some(br"\u0027"),
+            b'<' => Some(br"\u003c"),
+            b'>' => Some(br"\u003e"),
+            _ => None,
+        };
+        if let Some(escaped) = escaped {
+            f.write_str(unsafe { str::from_utf8_unchecked(&bytes[last..index]) })?;
+            f.write_str(unsafe { str::from_utf8_unchecked(escaped) })?;
```

Comment:
```suggestion
        let escaped = match byte {
            b'&' => Some(r"\u0026"),
            b'\'' => Some(r"\u0027"),
            b'<' => Some(r"\u003c"),
            b'>' => Some(r"\u003e"),
            _ => None,
        };
        if let Some(escaped) = escaped {
            f.write_str(unsafe { str::from_utf8_unchecked(&bytes[last..index]) })?;
            f.write_str(escaped)?;
```

We can at least remove one `unsafe` :)

Reply:
I copied this code verbatim from the current implementation. The rationale is that you cannot have a sized `str`, so `r"\u0026"` will load two words instead of just one.

I think the current implementation is "good enough" until [`#![feature(ascii_char)`](https://doc.rust-lang.org/unstable-book/library-features/ascii-char.html) is stable. Then we can strip one `unsafe` instruction without making the escaper slower. :)

Reply:
Maybe it's because it's late and I'm tired. But can you elaborate on the difference? I just glanced through the generated assembly, between using `r"\u0026"` vs `br"\u0026"`, and they result in bytes? Also since `&str` is essentially _just_ a fancy `&[u8]` _anyways_. What am I missing? :)

Reply:
Please have a look at <https://github.com/djc/askama/pull/805>. It can be that the compiler got better at optimizing in the meantime, though. I did not compare the generated code.

But I think changing that the implementation should be done in another PR, what do you think?

Reply:
Ah, that is interesting. I compared the assembly of before and after #805, along with a quick third variant I just wrote.

I definitely do see some differences. I do notice the increase of quad-words you mentioned before. I wonder if the compiler is simply better at optimizing a byte array. Whereas is seems it is less likely to treat a (even static) string as a byte array.

All in all, very interesting. But yes I agree, this is for a future separate PR.